rfd/0163-vnet.md

```diff
@@ -0,0 +1,971 @@
+outgoing TCP connections that have been made to Teleport apps, and DNS answers.
+
+#### DNS rebinding
+
+Because VNet will expose Teleport protected apps on the local machine over HTTP
+without TLS or any per-connection authentication, we need to be careful that
+they cannot be reached by untrusted code.
+
+One source of untrusted code that often runs on a user's machine is Javascript
+executed in the web browser.
+Web browsers apply CORS protections to ensure that Javascript cannot access
+arbitrary web addresses outside the domain that the JS code originated from.
+However, CORS does not protect plain HTTP websites from DNS rebinding attacks,
+where the attacker's website dynamically changes its IP address to the address
+of a website it wishes to make illegitimate HTTP requests to.
+
+Here's an example scenario:
+
+- The HTTP app `api.teleport.example.com` is listening on the VNet address `100.64.0.2`.
+- The user clicks a link to `hacker.example.com` which first resolves to a
+  public IP address with a very short TTL.
+- The web browser loads and executes the Javascript from `hacker.example.com`.
+- `hacker.example.com` "rebinds" its DNS address to `100.64.0.2`.
+- The Javascript for `hacker.example.com` makes a POST request to
+  `http://hacker.example.com/path`.
+- The browser will resolve the address of `hacker.example.com` to `100.64.0.2`
+  and the POST request will actually be sent to the `api.teleport.example.com`
+  app, authenticated as the user!
+
+There are a number of possible mitigations for this DNS rebinding attack
+scenario.
+
+1. Apply `Host` header allowlists for HTTP apps.
+2. Introduce TLS between clients and VNet for HTTP apps.
+3. Move HTTP app access out of scope for VNet (only provide access to TCP apps).
+4. Make VNet the global DNS nameserver on the host and reject any forwarded
+   queries that resolve to a VNet address or CNAME for a VNet name.
+5. Use a private IP range for VNet IPv4 addresses.
+6. Drop IPv4 and only support IPv6.
+
+`Host` header allowlists for HTTP apps are easy to implement, we will do it.
+This means that all HTTP apps will only be allowed to be accessed with a `Host`
+header exactly matching the registered app's `spec.public_addr`.
+Javascript loaded from `hacker.example.com` would have a mismatched `Host`
+header and would not be allowed to access any VNet HTTP apps.
+This should not impact most workflows.
+Unfortunately it will not protect any HTTP server actually registered as a TCP
+app in Teleport because VNet won't be able to assume that all requests can be
+parsed as HTTP.
+
+Introducing TLS between clients and VNet for HTTP apps would be tricky and
+wouldn't really give us any extra protection against DNS rebinding than `Host`
+header allowlists.
+The only reason HTTPS apps are generally safe from DNS rebinding attacks is that
+the `Host` in the request does not match any valid names for the certificate.
+The only way I think we could implement it would be to have VNet create a
```

Comment:
Brainstorming a couple alternative ways to introduce certificate validation:
1. Trust a CA instead of the individual self-signed cert.  This would work with custom DNS zones as well, but this is not something I want to recommend because of the security compromise it introduces too.
2. My other idea is more complicated, but could potentially leverage lets encrypt to get certificates for our internal subdomains.  Assuming a publicly available proxy which has vnet hosts routed to it (possibly using a wildcard dns record), it potentially could handle the let's encrypt validation of dns names needed by the vnet.

This would not be trivial, it also adds a significant complexity that vnet dns names could potentially be resolved by public DNS.  So we would need to ensure our resolver takes priority.  Since we are already requiring privileged access this seems potentially possible.

But the addition of validated https is a significant security gain too.

Reply:
to address point 1 (trust a CA instead of the individual self-signed cert), I think that's already what I have written here, I mention a self-signed CA trusted on the host, but also that I don't think it's worth doing

I don't really understand your second suggestion here to be honest. VNet is meant to run 100% locally on an end-user laptop, there's not meant to be a publicly available proxy involved. If we would be terminating TLS locally on the user device, we would need the private key for any trusted cert there, so it should not be any publicly trusted Let's Encrypt cert.

Can you help me understand the security gains you're alluding to of using validated HTTPS between the client and the local VNet service?

Reply:
Perhaps I misunderstood what you meant by trusting certs, I thought you meant on an individual self signed basis, where I mean a CA trust relationship.  But in any case, let's focus on point 2 since that is potentially an easier to adopt option for the customer.

> there's not meant to be a publicly available proxy involved

Using lets encrypt would require a publicly available proxy to handle the validation.  This may be an improvement we simply can't make, but having TLS even within the vnet would be valuable.  So I wanted to try and brainstorm possible options.

> Can you help me understand the security gains you're alluding to of using validated HTTPS between the client and the local VNet service?

Simply, it helps validate the authenticity of the endpoint we are connecting to.  We have to consider possible rebinding done by Teleport, but also possible rebinding done by other processes on the computer or the network.  Cert validation can help the user be confident they are talking to the proxy (and trust the proxy has done its own validation).  Without TLS cert validation, the user has no guarantee what they are connecting to (from the client).

Reply:
Oh I have an idea for this where VNet doesn't have to terminate TLS and the client can speak HTTPS...

Here's my summary of the situation: the client wants to do `curl https://restapi.teleport.example.com`, the teleport proxy has a TLS cert valid for `restapi.teleport.example.com` (as it must for app access to work) but requires the client to authenticate itself, and the client doesn't want to provide authentication

So VNet intercepts the TLS connection to `restapi.teleport.example.com` by pointing the DNS to one of its own addresses. VNet then forwards the client's TLS connection to the teleport proxy by wrapping it in an mTLS tunnel where VNet supplies client mTLS certs. The teleport proxy terminates the wrapping mTLS connection with VNet, and the inner TLS (HTTPS) connection with the client. The client is happy that it is speaking with `restapi.teleport.example.com`, and the proxy has authenticated the user

I think that would work well for HTTP(S) apps. Does this make sense? What do you think of the idea? It would require adding support to the teleport proxy

The thing is that for TCP apps, we can't do this, we need to forward TCP directly to the target app, the client may not speak TLS, or it might but that will be terminated by the target app

Reply:
^ This idea doesn't really work because TLS is often actually terminated by a load balancer in front of the Teleport proxy, so we would not be able to open a tunnel to the proxy and then have the proxy terminate TLS in that case.

DNS rebinding may be less of an issue now that we are not exposing web apps locally by default and only supporting TCP apps. The only risk is if a customer registers a plain HTTP app as a TCP app